util/route/route.go

```diff
@@ -0,0 +1,98 @@
+package route
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/net/context"
+)
+
+var (
+	mtx   = sync.RWMutex{}
+	ctxts = map[*http.Request]context.Context{}
+)
+
+// Context returns the context for the request.
+func Context(r *http.Request) context.Context {
+	mtx.RLock()
+	defer mtx.RUnlock()
+	return ctxts[r]
+}
+
+type param string
+
+// Param returns param p for the context.
+func Param(ctx context.Context, p string) string {
+	return ctx.Value(param(p)).(string)
+}
+
+// handle turns a Handle into httprouter.Handle
+func handle(h http.HandlerFunc) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		for _, p := range params {
```

Comment:
Yeah, that was a different situation in pushgateway. See for example: https://github.com/prometheus/pushgateway/blob/master/handler/delete.go#L33-L62

Params had to be protected there because the returned function closed over it, but it was passed in from the outside of the function constructor function (or whatever you want to call that :)).